### PR TITLE
Non-Record: McGilchrist Register Token — causal cumulative mean + FiLM global context pathway

### PR DESCRIPTION
**Pull request by aramdov**
## Method

Adds a **right-hemisphere global context pathway** to each transformer block, inspired by Iain McGilchrist's hemispheric asymmetry thesis (*The Master and His Emissary*).

Standard causal attention is a left-hemisphere machine: it decomposes inputs into local features, attends to specific positions, and synthesizes sequentially. What's missing is a **global context signal** — the right hemisphere's holistic, peripheral awareness of the whole document.

### Architecture

After the standard attention + MLP in each block:

```
Block output x: (B, T, D=512)
        │
        ▼
  cumsum(x, dim=1) / positions    ← causal cumulative mean (bf16, O(T) parallel)
        │
        ▼
  RMSNorm → Linear(D→8) → ReLU → Linear(8→2D)   ← bottleneck projection
        │
        ▼
  chunk → gamma, beta            ← FiLM parameters
        │
        ▼
  x_out = x * (1 + scale*gamma) + scale*beta     ← modulate block output
```

- **Fully causal**: position t only sees 0..t
- **O(T) parallel**: torch.cumsum, no sequential bottleneck  
- **Zero-initialized**: film_up weights = 0, register_scale = 0.01 → starts identical to SOTA
- **~123K extra params** at float16 (quantizer passthrough, compresses well)

### Results (v3, 2026-03-28)

| Metric | Value |
|--------|-------|
| val_bpb (int8+zstd) | **1.16465** |
| Training steps | 4,505 |
| Step time | 133ms |
| Peak VRAM | 24.9 GB / 80 GB |
| Eval time | 99s |
| Artifact | 16,855,808 bytes (**855KB over 16MB limit**) |

**Note on artifact size:** The 16MB overage comes from an untied output embedding head added in this run (`TIE_EMBEDDINGS=0`). The register mechanism itself adds only ~150KB compressed. v4 (remove untied head) is expected to come in under 16MB.

Per-step mechanism confirmed working — the register pathway learns and doesn't destabilize training.

### Infrastructure notes (for anyone building on this)

- `pip install zstandard` is required on every pod cold start (ephemeral container filesystem)
- `PYTORCH_ALLOC_CONF=expandable_segments:True` (note: `PYTORCH_CUDA_ALLOC_CONF` is deprecated)
- Batch size matters for step count: 2× batch tokens → 2× step time → same total tokens but half the gradient updates (bad for Muon)

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Non-Record: McGilchrist Register Token — causal cumulative mean + FiLM global context pathway

**Compliance:** LOOKS CLEAN — legal score-first-per-chunk TTT (PR #1413 pattern)

LoRA TTT per-doc reset, BigramHash legal. Clean.

**Verdict:** LOOKS CLEAN — legal TTT implementation matching the PR #1413 (dexhunter) pattern: each chunk scored under `torch.no_grad()` before `optimizer.step()`, with `is_last_chunk` guard preventing adaptation on the final scored chunk.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). TTT implementation follows the legal score-first discipline.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source, cross-checked against deterministic AST classifier. If this review misread your code, please call it out so I can re-audit manually.*
